relaynet-cogrpc-jvm/src/main/java/tech/relaycorp/relaynet/cogrpc/client/CogRPCClient.kt

```diff
@@ -1,19 +1,132 @@
+class CogRPCClient(
+    serverAddress: String,
+    useTls: Boolean = true
+) {
+
+    private val channel by lazy {
+        val url = URL(serverAddress)
+        val address = InetSocketAddress(
+            url.host,
+            url.port.let { if (it != -1) it else DEFAULT_PORT }
+        )
+        NettyChannelBuilder
+            .forAddress(address)
+            .useTransportSecurity()
+            .run { if (useTls) useTransportSecurity() else usePlaintext() }
+            .build()
+    }
 
     @Throws(Exception::class)
-    fun deliverCargo(cargoes: Iterable<CogRPC.MessageDelivery>): Flow<CogRPC.MessageDeliveryAck>
+    fun deliverCargo(cargoes: Iterable<CogRPC.MessageDelivery>): Flow<CogRPC.MessageDeliveryAck> {
+        if (cargoes.none()) return emptyFlow()
+
+        val cargoesToAck = cargoes.map { it.localId }.toMutableList()
+        val ackChannel = BroadcastChannel<CogRPC.MessageDeliveryAck>(1)
+        val ackObserver = object : StreamObserver<CargoDeliveryAck> {
+            override fun onNext(value: CargoDeliveryAck) {
+                logger.info("deliverCargo ack ${value.id}")
+                ackChannel.sendBlocking(CogRPC.MessageDeliveryAck(value))
+                cargoesToAck.remove(value.id)
+                if (cargoesToAck.isEmpty()) {
+                    logger.info("deliverCargo ack complete")
+                    ackChannel.close()
+                }
+            }
+
+            override fun onError(t: Throwable) {
+                logger.log(Level.WARNING, "deliverCargo ack error", t)
+                ackChannel.close(t)
+            }
+
+            override fun onCompleted() {
+                logger.info("deliverCargo ack closed")
+                ackChannel.close()
+            }
+        }
+
+        val client = buildClient()
+        val deliveryObserver = client.deliverCargo(ackObserver)
+
+        cargoes.forEach { delivery ->
+            logger.info("deliverCargo next ${delivery.localId}")
+            deliveryObserver.onNext(delivery.toCargoDelivery())
+        }
+        logger.info("deliverCargo complete")
+        deliveryObserver.onCompleted()
+
+        return ackChannel.asFlow()
+    }
 
     @Throws(Exception::class)
-    fun collectCargo(cca: CogRPC.MessageDelivery): Flow<CogRPC.MessageReceived>
+    fun collectCargo(cca: CogRPC.MessageDelivery): Flow<CogRPC.MessageDelivery> {
+        val ackChannel = BroadcastChannel<CogRPC.MessageDeliveryAck>(1)
+        return channelFlow {
+            val collectObserver = object : StreamObserver<CargoDelivery> {
+                override fun onNext(value: CargoDelivery) {
+                    logger.info("collectCargo ${value.id}")
+                    this@channelFlow.sendBlocking(CogRPC.MessageDelivery(value))
+                    ackChannel.sendBlocking(CogRPC.MessageDeliveryAck(value.id))
+                }
+
+                override fun onError(t: Throwable) {
+                    logger.log(Level.WARNING, "collectCargo error", t)
```

Comment:
Per our convo yesterday, I think the courier app should deleted the CCA if the gRPC status code is `PERMISSION_DENIED`.

From the point of view of this client, I guess one way to support that is to create a new exception class (e.g., `CCARefusedError`) and throw it here when the server ends the call with a `PERMISSION_DENIED` error. Wdyt?